### ISSUE TITLE
@Context() in services

### ISSUE THREAD

**Issue by luca-nardelli** [doug-martin/nestjs-query#861]
**Is your feature request related to a problem? Please describe.**
I'm currently facing the following situation:
- A user has a "customer" profile (i.e. User entity only holds auth information, Customer holds all the other relevant data), they are related in a OneToOne fashion.
- Customers can create and own (i.e. database OneToMany relation) `projects`
- When creating a project, I'd like to automatically set its customer property to the customer identified by the logged in user
- My AuthGuard only loads the User entity (without Customer entity) to avoid unnecessary reads from the db.

I already have the user from the Authentication guards (in req.user). In order to set the project owner, I would need need to fetch the customer entity from the db given the user, and then assign it to the project that's being created. However, I couldn't find a way to do that "cleanly".
In hooks (BeforeCreateOne), I have access to the context, however the user loaded in the context does not have customer information attached and hooks cannot access other services since they are plain functions. I could load the customer profile in the guard, but it seems a waste of resources to load it on every request even if it's not needed (e.g. in queries or other mutations).
In the QueryService (which seems to me the right place where this operation should happen, as it is business logic), I don't have access to the context, as it is not passed by the CRUDResolver. If I had the context, I could simply extend the QueryService, load the customer and assign it to the project being created.

### Have you read the [Contributing Guidelines](https://github.com/doug-martin/nestjs-query/blob/master/CONTRIBUTING.md)?

Yes.

**Describe the solution you'd like**
Having access to the context in the QueryService would be great, so that we can change the behavior of the service based on its properties. All methods in the QueryServices (createOne, createMany, updateOne, etc...) should accept an additional parameter which is the context that can be injected in the resolver via @Context() decorator.

**Describe alternatives you've considered**
The only alternative I have so far is to force the frontend to send the "customer" parameter in the mutation even if it could be inferred by the server given the logged in user.

**Additional context**
If the above description is not clear, I can try to write down something more detailed.

I think I could work on a PR for this, as I have digged through the sources a bit trying to find a solution for this issue.

**Comment by doug-martin** [comment]
@luca-nardelli Off the top of my head I can think of a couple of solutions that may work out of the box

* [Inject the context or request into your service](https://docs.nestjs.com/fundamentals/injection-scopes#request-provider). 
* Create a guard that adds the customer information to the context and only add it to your create one/many mutations.

Let me know if that helps, we do something similar internally and manage it through guards to check that the requesting user has access to a certain entity and add it to the context to be used in a hook or authorizer.

**Comment by smolinari** [comment]
I think there is another choice. You could use Nest's caching module. Cache what you need at the point of creating the user's session (so only one call to the db for this "static" data) and recall it into any custom service via Nest's DI system.

Scott

**Comment by luca-nardelli** [comment]
Hey, thank you both for the suggestions! I come from PHP and and I probably still need to get used to the subtle differences =). I thought about using a request scoped provider (and I forgot to write it in the initial description, sorry :/) to inject the request in the service, but if I understood correctly how that works then Nest would bubble up the scope to the resolver, re-instantiating it for each request, and that seems an unnecessary performance hit. 

As for the guard, that makes sense, however this doesn't seem to be the correct purpose of a guard, is it? Ideally I'd like my services to perform all the business logic related to the resource (something similar to what [FeathersJS](https://docs.feathersjs.com/api/services.html) does) given a specific input + context, so that, for example, I can re-use the same services in CLI commands (or in REST/RPC endpoints, for example) without having to change much. By using the guard I would have to remember to decorate all the other controllers/handlers, and I don't think I could use it for CLI commands.

As for the cache, that could also be an option, but without the request object in the service I think wouldn't know how to compute the cache key to retrieve the data, or am I missing something?

In any case, for my immediate issue I'll follow the guard + hook approach, but if you think what I said above about the service incorporating business logic makes some sense, let's keep the discussion open! I could always create a `ContextAwareCRUDResolver` and `ContextAwareQueryService` classes for my use case, but maybe that's a completely wrong approach.

**Comment by smolinari** [comment]
> As for the cache, that could also be an option, but without the request object in the service I think wouldn't know how to compute the cache key to retrieve the data, or am I missing something?

Your right... I forgot to mention we are using the convention of passing the context from the resolvers through to our services (manual injection). So, any service will have access to the request object and the user info stored within the access token. From there, you can pull in the cached user data. 

If there is a more elegant way than doing that, I'm all ears too. 😁 

Scott

**Comment by doug-martin** [comment]
@luca-nardelli in the latest version (v0.23.0) I have added the ability to use DI in hooks, which should provide a clean place to put this type of logic. 

Check out the [release notes](https://github.com/doug-martin/nestjs-query/discussions/911) and [the hooks documentation](https://doug-martin.github.io/nestjs-query/docs/graphql/hooks)

Let me know what you think! Thanks again for the feedback.

**Comment by luca-nardelli** [comment]
Yes it totally makes sense! Thanks for the release! I would still love to have access to the context in services, so that if in the future I have to implement a REST endpoint on top of that I would not need to remember to add the hooks there as well because the logic is in the service itself, but that's a design that probably requires further discussion. Thanks again for the release! Should we close this?

**Comment by luca-nardelli** [comment]
Hey! I'm posting here since we started discussing the topic just to inform you about a package I just found that is using Node async_hooks to provide per request context out of the box. I haven't tried it at the moment but 
I believe this could be easily used to achieve what I needed at the time I opened this issue! I'll update the issue once/if I try this

Package: https://www.npmjs.com/package/@medibloc/nestjs-request-context
